java/src/main/native/src/RmmJni.cpp

```diff
@@ -79,18 +90,54 @@ class tracking_resource_adaptor final : public base_tracking_resource_adaptor {
 
   std::size_t get_total_allocated() override { return total_allocated.load(); }
 
+  void push_thread_memory_tracker() override { memory_tracker_stack.emplace(); }
+
+  long pop_thread_memory_tracker() override {
+    auto top_tracker = memory_tracker_stack.top();
+    auto ret = top_tracker.max_outstanding;
+    memory_tracker_stack.pop();
+    if (memory_tracker_stack.empty()) {
+      alloc_map.clear();
+    } else {
+      // carry the max to the next level
+      memory_tracker_stack.top().max_outstanding += ret;
+    }
+    return ret;
+  }
+
 private:
   Upstream *const resource;
   std::size_t const size_align;
   std::atomic_size_t total_allocated{0};
 
+  void thread_allocated(long addr, std::size_t num_bytes) {
+    if (!memory_tracker_stack.empty()) {
+      alloc_map[addr] = num_bytes;
+      memory_tracker &tracker = memory_tracker_stack.top();
+      tracker.current_outstanding += num_bytes;
+      tracker.max_outstanding = std::max(tracker.current_outstanding, tracker.max_outstanding);
+    }
+  }
+
+  void thread_freed(long addr, std::size_t num_bytes) {
+    if (!memory_tracker_stack.empty()) {
+      auto it = alloc_map.find(addr);
+      if (it != alloc_map.end()) {
+        auto tracker = memory_tracker_stack.top();
+        tracker.current_outstanding -= it->second;
+        alloc_map.erase(it);
+      }
+    }
+  }
```

Comment:
It feels like there is a bug or unintended behavior here. Consider:

```
tracking_resource_adapter mr{...};

mr.push_thread_memory_tracker(); // stack[0]
auto p0 = mr.allocate(...);
mr.push_thread_memory_tracker(); // stack[1]
mr.deallocate(p0); // bug(?) this will cause stack[1].current_outstanding to go negative
```

In other words, because the `alloc_map` spans all push/pop scopes, if you deallocate a pointer from an outer scope inside a nested scope, you'll end up messing up your accounting.

Reply:
So we are going in a different direction. This code doesn't exist anymore :)